src/Core/IdentityServer/ResourceOwnerPasswordValidator.cs

```diff
@@ -60,25 +60,32 @@ public async Task ValidateAsync(ResourceOwnerPasswordValidationContext context)
             //    return;
             //}
 
-            if (_captchaValidationService.ServiceEnabled && _currentContext.IsBot)
+            await ValidateAsync(context, context.Request);
+
+            if (_captchaValidationService.ServiceEnabled && (_currentContext.IsBot || _captchaValidationService.RequireCaptcha))
             {
-                var captchaResponse = context.Request.Raw["CaptchaResponse"]?.ToString();
-                if (string.IsNullOrWhiteSpace(captchaResponse))
+                var twoFactorVerified = context.Result.CustomResponse.ContainsKey("TwoFactorVerified") ?
+                    (bool)context.Result.CustomResponse["TwoFactorVerified"] :
+                    false;
```

Comment:
This seems kind of hacky. I guess this got set in `ValidateAsync` above?

Reply:
it's set for successful responses, otherwise this sets it to false.

It is a little weird, but the alternative is an `out` parameter or a tuple, which felt a little extreme